File: Server/Modules/Utils/DAL/Common/Entities/EntitySet.ts

import Entity = require("./Entity");

class EntitySet<T extends IEntity> implements IEntitySet<T> {
    constructor(public entityTypeName: string, private entitySets: IDictionary<IEntitySet<any>>, private metadata: metadataTypes.Metadata) {
        this.items = [];
        this.key = this.metadata.entityTypes[this.entityTypeName].key;
        this.EntityClass = this.createEntityClass(this.metadata);
    }

    private items: T[];
    private key: string[];
    public EntityClass: any;

    navigateSingle(remoteEntity: any, remoteEntityKey: string[], navigationKey: string[]): T {
        var result = this.find((it) => this.haveSameKeysNavigation(it, navigationKey, remoteEntity, remoteEntityKey));
        return result;
    }

    navigateMulti(remoteEntity: any, remoteEntityKey: string[], navigationKey: string[]): T[] {
        var result = this.filter((it) => this.haveSameKeysNavigation(it, navigationKey, remoteEntity, remoteEntityKey));
        return result;
    }

    findByKey(partialEntity: any): T {
        var result = this.find((it) => this.haveSameKeysLocal(it, partialEntity));
        return result;
    }

    find(predicate: IPredicate<T>): T {
        for (let item of this.items) {
            if (predicate(item)) {
                return item;
            }
        }
        return null;
    }

    filter(predicate: IPredicate<T>): T[] {
        return this.items.filter(predicate);
    }

    createEntity(): T {
        var entity = new this.EntityClass();
        return entity;
    }

    deleteEntity(entity: T): void {
        var entityIndex = this.items.indexOf(entity);
        if (entityIndex !== -1) {
            entity["_entitySets"] = undefined;
            this.items.splice(entityIndex, 1);
        }
    }

    deleteAll(): void {
        for (let item of this.items) {
            item["_entitySets"] = undefined;
        }
        this.items.splice(0, this.items.length);
    }

    dispose(): void {
        this.deleteAll;
        this.entitySets = undefined;
        this.metadata = undefined;
    }

    updateEntity(dto: any): T {

        var found: any, foundIndex: number, newItem: T;

        // se cauta elementul in colectia existenta
        found = this.findByKey(dto);
        if (found === null) {
            // daca nu a fost gasit se adauga in colectie
            newItem = this.createNewItem(dto);
            this.items.push(newItem);
        } else {
            // daca a fost gasit nu se inlocuieste ci se actualizaeza datale
            // astfel ca astfel ca referintele din dataViews existente sa nu se piarda.
            newItem = this.initialize(dto, found);
        }

        return newItem;
    }

    private createNewItem(dto: any): T {
        var entity: T = this.createEntity();
        //entity["_entitySets"] = this.entitySets;
        entity._attach(this.entitySets);
        this.initialize(dto, entity);
        return entity;
    }

    private initialize(dto: any, entity: T): T {
        for (let prop in entity) {
            if (prop !== "_entitySets" && dto.hasOwnProperty(prop)) {
                if (dto[prop] && this.metadata.entityTypes[this.entityTypeName].properties[prop].type === "Date") {
                    entity[prop] = new Date(dto[prop]);
                } else {
                    entity[prop] = dto[prop];
                }
            }
        }
        return entity;
    }

    private haveSameKeysLocal(localEntity: any, remoteEntity: any): boolean {
        for (let i = 0; i < this.key.length; i++) {
            if (remoteEntity[this.key[i]] !== localEntity[this.key[i]]) {
                return false;
            }
        }
        return true;
    }

    private haveSameKeysNavigation(localEntity: any, keyLocal: string[], remoteEntity: any, keyRemote: string[]): boolean {
        for (let i = 0; i < keyLocal.length; i++) {
            if (remoteEntity[keyRemote[i]] !== localEntity[keyLocal[i]]) {
                return false;
            }
        }
        return true;
    }

    private createEntityClass(metadata: metadataTypes.Metadata): any {
        // initialize EntityClass properties
        var properties = metadata.entityTypes[this.entityTypeName].properties;
        var entityClass = class extends Entity {
            constructor() {
                super();
                var self: any = this;
                for (let prop in properties) {
                    self[prop] = properties[prop].default;
                }
            }
        };

        // initialize EntityClass navigation properties
        var entityType = metadata.entityTypes[this.entityTypeName];
        for (let navigationPropertyName in entityType.navigationProperties) {

            var metaNavElement = entityType.navigationProperties[navigationPropertyName];

            var navProp = (function (metaNavElement: metadataTypes.NavigationProperty) {
                var _entityTypeName = metaNavElement.entityTypeName;
                var _multiplicity = metaNavElement.multiplicity;
                var _keyLocal = metaNavElement.keyLocal;
                var _keyRemote = metaNavElement.keyRemote;
                var result = function () {
                    // se verifica daca EntitySet-ul este creeat in dataSet
                    // (intrucat acestea se creeaza doar daca este necesar - Lazy)
                    var remoteEntitySet: IEntitySet<any> = this._entitySets.hasOwnProperty(_entityTypeName) ? this._entitySets[_entityTypeName] : undefined;
                    if (_multiplicity === "multi") {
                        return remoteEntitySet ? remoteEntitySet.navigateMulti(this, _keyLocal, _keyRemote) : [];
                    } else {
                        return remoteEntitySet ? remoteEntitySet.navigateSingle(this, _keyLocal, _keyRemote) : null;
                    }
                };
                return result;
            })(metaNavElement);

            //// proprietati de navigare ca functii
            //metadata[entityName].prototype[navigationProperty] = navProp;

            // proprietati de navigare ca get-ers
            Object.defineProperty(entityClass.prototype, navigationPropertyName, {
                get: navProp,
                enumerable: false,
                configurable: true
            });

        }

        return entityClass;
    }

}

export = EntitySet;

File: Server/Models/Utils/DAL/Common/Entities/EntitySet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace Server.Models.Utils.DAL.Common
{
    public class EntitySet<T> : IEntitySet<T>
        where T : class, IEntity
    {
        public EntitySet(Type entityType, Dictionary<string, IEntitySet<IEntity>> entitySets, Metadata metadata)
        {
            this.entityTypeName = entityType.Name;
            this.entitySets = entitySets;
            this.metadata = metadata;
            this.key = metadata.EntityTypes[this.entityTypeName].Key;
            this.localKeyPropertyInfo = entityType.GetProperties().Where(prop => this.key.Contains(prop.Name)).ToArray();
            this.Items = new List<T>();
        }

        private readonly string entityTypeName;
        private Dictionary<string, IEntitySet<IEntity>> entitySets;
        private Metadata metadata;

        private readonly string[] key;
        private readonly PropertyInfo[] localKeyPropertyInfo;

        public List<T> Items { get; private set; }

        //public static IEntitySet<IEntity> CreateEntitySet(Dictionary<string, IEntitySet<IEntity>> entitySets, Metadata metadata)
        //{
        //    var d1 = typeof(EntitySet<>);
        //    var typeArgs = new Type[] { typeof(IEntity) };
        //    var constructed = d1.MakeGenericType(typeArgs);
        //    var entitySet = (IEntitySet<IEntity>)Activator.CreateInstance(constructed, new object[] { entitySets, metadata });
        //    return entitySet;
        //}

        public T NavigateSingle(IEntity remoteEntity, string[] remoteEntityKey, string[] navigationKey)
        {
            var result = this.Items.FirstOrDefault((it) => this.HaveSameKeysNavigation(it as Dto, navigationKey, remoteEntity as Dto, remoteEntityKey));
            return result;
        }

        public IEnumerable<T> NavigateMulti(IEntity remoteEntity, string[] remoteEntityKey, string[] navigationKey)
        {
            var result = this.Items.Where((it) => this.HaveSameKeysNavigation(it as Dto, navigationKey, remoteEntity as Dto, remoteEntityKey));
            return result;
        }

        public IEnumerable<T> NavigateAllRelated(IEnumerable<object> remoteEntities, string[] remoteEntityKey, string[] navigationKey)
        {
            var result = this.Items.Where((it) =>
            {
                foreach (var remoteEntity in remoteEntities)
                {
                    if (this.HaveSameKeysNavigation(it as Dto, navigationKey, (Dto)remoteEntity, remoteEntityKey))
                    {
                        return true;
                    }
                }
                return false;
            });
            return result;
        }

        public T FindByKey(T partialEntity)
        {
            var result = this.Items.FirstOrDefault((it) => this.HaveSameKeysLocal(it as Dto, partialEntity as Dto));
            return result;
        }

        public T Find(Func<T, bool> predicate)
        {
            var result = this.Items.FirstOrDefault(predicate);
            return result;
        }

        public IEnumerable<T> Filter(Func<T, bool> predicate)
        {
            var result = this.Items.Where(predicate);
            return result;
        }

        public void DeleteEntity(T entity)
        {
            entity._detach();
            this.Items.Remove(entity);
        }

        public void DeleteAll()
        {
            foreach (var entity in this.Items)
            {
                entity._detach();
            }
            this.Items.RemoveAll((it) => true);
        }

        public void Dispose()
        {
            this.DeleteAll();
            this.entitySets = null;
            this.metadata = null;
        }

        public T UpdateEntity(T dto)
        {
            T newItem;
            // se cauta elementul in colectia existenta
            var found = this.FindByKey(dto);
            if (found == null)
            {
                // daca nu a fost gasit se adauga in colectie
                newItem = this.CreateNewItem(dto);
                this.Items.Add(newItem);
            }
            else
            {
                // daca a fost gasit nu se inlocuieste ci se actualizaeza datale
                // astfel ca astfel ca referintele din dataViews existente sa nu se piarda.
                newItem = this.Initialize(dto as Dto, found as Dto);

            }

            return newItem;
        }

        public void AttachEntitySet(List<T> dtos)
        {
            foreach (var dto in dtos)
            {
                this.CreateNewItem(dto);
            }
            this.Items = dtos as List<T>;
        }

        private T CreateNewItem(T dto)
        {
            dto._attach(this.entitySets, this.metadata);
            return dto;
        }

        private T Initialize(Dto dto, Dto entity)
        {
            foreach (var prop in dto)
            {
                entity[prop.Key] = prop.Value;
            }
            return entity as T;
        }

        private bool HaveSameKeysLocal(Dto localEntity, Dto remoteEntity)
        {
            for (int i = 0; i < this.key.Length; i++)
            {
                if ((int)localEntity[this.key[i]] != (int)remoteEntity[this.key[i]])
                {
                    return false;
                }
            }
            return true;
        }

        private bool HaveSameKeysNavigation(Dto localEntity, string[] keyLocal, Dto remoteEntity, string[] keyRemote)
        {
            for (int i = 0; i < keyLocal.Length; i++)
            {
                if ((int)localEntity[keyLocal[i]] != (int)remoteEntity[keyRemote[i]])
                {
                    return false;
                }
            }
            return true;
        }
    }

}